pkg/spoke/statusfeedback/reader.go

```diff
@@ -0,0 +1,121 @@
+package statusfeedback
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/util/jsonpath"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+	"open-cluster-management.io/work/pkg/spoke/statusfeedback/rules"
+)
+
+type StatusReader struct {
+	wellKnownStatus rules.WellKnownStatusRuleResolver
+}
+
+func NewStatusReader() *StatusReader {
+	return &StatusReader{
+		wellKnownStatus: rules.DefaultWellKnownStatusRule(),
+	}
+}
+
+func (s *StatusReader) GetValuesByRule(obj *unstructured.Unstructured, rule workapiv1.FeedbackRule) ([]workapiv1.FeedbackValue, error) {
+	errs := []error{}
+	values := []workapiv1.FeedbackValue{}
+
+	switch rule.Type {
+	case workapiv1.WellKnownStatusType:
+		gvk := obj.GroupVersionKind()
+		paths := s.wellKnownStatus.GetPathsByKind(gvk.GroupKind())
+		if len(paths) == 0 {
+			break
+		}
+
+		for _, path := range paths {
+			value, err := getValueByJsonPath(path.Name, path.Path, obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if value == nil {
+				continue
+			}
+			values = append(values, *value)
+		}
+	case workapiv1.JSONPathsType:
+		for _, path := range rule.JsonPaths {
+			// skip if version is specified and the object version does not match
+			if len(path.Version) != 0 && obj.GroupVersionKind().Version != path.Version {
+				continue
+			}
+
+			value, err := getValueByJsonPath(path.Name, path.Path, obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if value == nil {
+				continue
+			}
+			values = append(values, *value)
+		}
+	}
+
+	return values, utilerrors.NewAggregate(errs)
+}
+
+func getValueByJsonPath(name, path string, obj *unstructured.Unstructured) (*workapiv1.FeedbackValue, error) {
+	j := jsonpath.New(name).AllowMissingKeys(true)
+	err := j.Parse(fmt.Sprintf("{.status%s}", path))
```

Comment:
implicitly adding `.status` as the prefix of the json path is suspicious. even tho it's always recommended to extend the status information under the `.status` struct but there can be other forms of saving status in the object such as annotations or even the spec.